Compression.md

```diff
@@ -0,0 +1,145 @@
+## WebSocket Compression
+WebSocket Compression, defined by [RFC7692](https://tools.ietf.org/html/rfc7692) allows WebSocket clients to send and receive compressed data on a WebSocket connection. Compression reduces the total wire-level payload of a WebSocket connection, possibly resulting in an improved throughput.
+
+This document discusses the implementation of WebSocket Compression in [Kitura-WebSocket-NIO](https://github.com/IBM-Swift/Kitura-WebSocket-NIO), which is an implementation of the [Kitura-WebSocket](https://github.com/IBM-Swift/Kitura-WebSocket) API using [swift-nio](https://github.com/apple/swift-nio).
+
+This document assumes the reader is aware of the fundamentals of the [WebSocket protocol](https://tools.ietf.org/html/rfc6455).
+
+### Table of contents
+1. [WebSocket extensions](https://gist.github.com/pushkarnk/aea1be88c7c7283fcfe4615df35ed857#1-websocket-extensions)
+2. [WebSocket compression - the `permessage-deflate` algorithm](https://gist.github.com/pushkarnk/aea1be88c7c7283fcfe4615df35ed857#2-the-permessage-deflate-algorithm)
+3. [An implementation of permessage-deflate based on swift-nio](https://gist.github.com/pushkarnk/aea1be88c7c7283fcfe4615df35ed857#3-an-implementation-of-permessage-deflate-based-on-swift-nio)
+4. [Developer notes](https://gist.github.com/pushkarnk/aea1be88c7c7283fcfe4615df35ed857#4-developer-notes)
+
+### 1. WebSocket extensions
+The WebSocket protocol has a provision for servers to configure protocol extensions, and for clients to request these extensions from the servers. A client notifies about extensions it is interested in through a `negotiation offer` using the `Sec-WebSocket-Extension` header. A server may or may not support the extensions requested by the client. Through a `negotiation response`, the server notifies the client of the extensions that the server agrees upon. The negotiation response offer and response may also include extension-specific parameters. Once an extension is agreed upon, the client and server must invoke the extension from their respective WebSocket implementations.
+
+WebSocket compression is a WebSocket extension.
+
+### 2. WebSocket Compression: the permessage-deflate algorithm
+Permessage-deflate is a WebSocket extension defined by [RFC7692](https://tools.ietf.org/html/rfc7692) which provides a specification for the compression functionality. It defines the negotiation process and a compression algorithm called `DEFLATE`. 
+
+A `permessage-deflate` negotiation happens during the upgrade request from HTTP to WebSocket. A `permessage-deflate` negotiation offer has a mandatory `permessage-deflate` string followed by a semi-colon separated list of extension parameters. There are four extension parameters defined for WebSocket compression:
```

Comment:
Given there is quite a lot of text in here it might make sense to have a "`permessage-deflate` negotiation offer" minor heading and "`permessage-deflate` negotiation response" minor heading to make this section a little easier to read. It's just a suggestion though.